airflow/operators/datetime_branch.py

```diff
@@ -0,0 +1,77 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import datetime
+from typing import Dict, Iterable, Optional, Union
+
+from airflow.exceptions import AirflowException
+from airflow.operators.branch_operator import BaseBranchOperator
+from airflow.utils import timezone
+from airflow.utils.decorators import apply_defaults
+
+
+class DateTimeBranchOperator(BaseBranchOperator):
+    """
+    Branches into one of two lists of tasks depending on the current datetime.
+
+    True branch will be returned when `datetime.datetime.now()` falls below
+    `target_upper` and above `target_lower`.
+
+    :param follow_task_ids_if_true: task id or task ids to follow if
+        `datetime.datetime.now()` falls above target_lower and below `target_upper`.
+    :type follow_task_ids_if_true: str or list[str]
+    :param follow_task_ids_if_false: task id or task ids to follow if
+        `datetime.datetime.now()` falls below target_lower or above `target_upper`.
+    :type follow_task_ids_if_false: str or list[str]
+    :param target_lower: target lower bound.
+    :type target_lower: Optional[datetime.datetime]
+    :param target_upper: target upper bound.
+    :type target_upper: Optional[datetime.datetime]
+    """
+
+    @apply_defaults
+    def __init__(
+        self,
+        *,
+        follow_task_ids_if_true: Union[str, Iterable[str]],
+        follow_task_ids_if_false: Union[str, Iterable[str]],
+        target_lower: Optional[datetime.datetime],
+        target_upper: Optional[datetime.datetime],
```

Comment:
I'm wondring about the use cases of datetime in general.
I would assume that it's more likely that user would like to branch based on times rather than datetime.
Do you think we will need a different TimeBranchOperator or the two use cases can be combined?

Reply:
Branching of times should be possible by setting the target values to something like:
```
target_lower=datetime.datetime.combine(
    datetime.datetime.now(), datetime.time(hour=target_hour, minute=target_minute, second=target_second)
)
```

I think adding a TimeBranchOperator may be a bit redundant, instead, we could support targets of type `Optional[Union[datetime.datetime, datetime.time]]`, and when encountering a `datetime.time` we combine it with the current date and proceed as it is right now:

```
def choose_branch(self, context: Dict) -> Union[str, Iterable[str]]:
    now = timezone.make_naive(timezone.utcnow(), self.dag.timezone)
    target_upper = datetime.datetime.combine(now, self.target_upper) if isinstance(self.target_upper, datetime.time) else self.target_upper
```

What do you think?

Reply:
yeah I think the `Optional[Union[datetime.datetime, datetime.time]]` approach is better as it simplify the usage for the users but maybe worth asking for more opinions on this one.

note that there are edge cases for example lower is 23:00 and upper is 01:00 (the next day) so the duration of success is total of 2 hours yet it spread over two dates.

Also it will be good to add [documentation](https://github.com/apache/airflow/tree/master/docs/howto/operator) about this operator. you can use previous PRs to see examples https://github.com/apache/airflow/pull/11472

Reply:
Figured out it may be easier to hear other opinions with a concrete example, so I implemented the idea mostly as described. The edge case you described is handled by pushing the smaller time to the next day. Should the same thing be done when both times are equal? Or should an exception be raised instead? 

Gave it a shot at adding documentation, do let me know if it needs to be extended or rewritten.